File: 2365-task-scheduler-ii/2365-task-scheduler-ii.py
class Solution:
    def taskSchedulerII(self, tasks: List[int], space: int) -> int:
        time = 0
        ptr = 0
        mp = {}
        
        while ptr<len(tasks):
            if tasks[ptr] in mp and mp[tasks[ptr]]>time:
                time = mp[tasks[ptr]]
            else:
                mp[tasks[ptr]] = time+space+1
                time+=1
                ptr+=1
        return time

File: 2365-task-scheduler-ii/2365-task-scheduler-ii.java
class Solution {
    public long taskSchedulerII(int[] tasks, int space) {
        long time = 0;
        Map<Integer, Long> map = new HashMap<>();
        
        for(int ptr=0; ptr<tasks.length; ++ptr){
            if(map.containsKey(tasks[ptr]) && map.get(tasks[ptr])>time){
                time = (long)map.get(tasks[ptr]);
                ptr--;
            }
            else{
                map.put(tasks[ptr], time+space+1);
                time++;
            }
        }
        
        return time;
    }
}